### PR TITLE
Upgrade OKEX API from v1 to v3

### PR DESCRIPTION
**Pull request by EdNoepel**
Key changes include order_id becoming a string and a change to the structure of account balances for each token.  Some fields were renamed (deal for example) to improve consistency with other exchanges and the new OKEX API.  An implementation of get_trades is now available, prepared using filled order information.  Pagination is broken on OKEX's side; cursor implementation is left commented out in case they eventually resolve the problem.  Unit tests which mock REST API responses are implemented.

This implementation is not backwards-compatible; pull requests for consuming components (including keeper) will be created shortly.

### CODE REVIEW

File: pyexchange/okex.py [hunk 1]
```diff
@@ -121,93 +126,158 @@ def __repr__(self):
...
                                             volume=Wad.from_number(item[5])), result))
 
+    # Account: Get available and frozen balances for each token
     def get_balances(self) -> dict:
-        return self._http_post("/api/v1/userinfo.do", {})["info"]["funds"]
+        result = self._http_get("/api/spot/v3/accounts", "", requires_auth=True)
 
-    def get_orders(self, pair: str) -> List[Order]:
-        assert(isinstance(pair, str))
+        balances = {}
```
[the remark this answers is not in the document]
Comment (EdNoepel) [reply pyexchange/okex.py]: The v1 implementation of _get_balances_ returned a dictionary keyed by balance type ("frozen", "available", and such), which came directly from the server.  Left unchanged, the v3 implementation would provide a (unindexed) list.  Indexing by currency preserves the return type and likely is most useful to consumers.  Dependent projects (_inventory service_) have been updated accordingly.

File: pyexchange/okex.py [hunk 2]
```diff
@@ -29,23 +32,25 @@
...
 class Order:
-    def __init__(self, order_id: int, timestamp: int, pair: str,
-                 is_sell: bool, price: Wad, amount: Wad, deal_amount: Wad):
-        assert(isinstance(order_id, int))
+    def __init__(self, order_id: str, timestamp: int, pair: str, status: str,
+                 is_sell: bool, price: Wad, amount: Wad, filled_amount: Wad):
+        assert(isinstance(order_id, str))
         assert(isinstance(timestamp, int))
         assert(isinstance(pair, str))
+        assert(isinstance(status, str))
```
[the remark this answers is not in the document]
Comment (EdNoepel) [reply pyexchange/okex.py]: I can remove this and update consuming projects accordingly.  Should I update *get_orders_history* to filter out filled orders?

File: pyexchange/okex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (grandizzy) [reply pyexchange/okex.py]: makes sense

File: pyexchange/okex.py [hunk 3]
```diff
@@ -121,93 +126,158 @@ def __repr__(self):
...
+        assert(isinstance(pair, str))
 
-    def get_orders_history(self, pair: str, number_of_orders: int) -> List[Order]:
+        result = self._http_get("/api/spot/v3/orders_pending",
+                                f"&instrument_id={pair}",
+                                requires_auth=True, has_cursor=False)
+
+        return list(map(self._parse_order, result))
+
+    # Trading: Retrieves order list for a particular pair, newest first.
```
[the remark this answers is not in the document]
Comment (EdNoepel) [reply pyexchange/okex.py]: I'll remove the commented-out implementation which uses cursors for pagination.  If OKEX ever fixes their implementation, we can recover that work to return more order history.

File: pyexchange/okex.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (grandizzy) [reply pyexchange/okex.py]: yes, or at least make it optional or put default value

File: pyexchange/okex.py [hunk 4]
```diff
@@ -217,118 +287,190 @@ def place_order(self, pair: str, is_sell: bool, price: Wad, amount: Wad) -> int:
...
+            'type': 'limit',
+            'side': 'sell' if is_sell else 'buy',
             'price': float(price),
-            'amount': float(amount)
+            'size': float(amount),
+            'margin': 0
         })
-        order_id = int(result['order_id'])
+        print(result)
+        order_id = result['order_id']
```
[the remark this answers is not in the document]
Comment (EdNoepel) [reply pyexchange/okex.py]: In v3, *order_id* and *trade_id* become strings.  Although testing reveals the strings return only numeric data, I believe I should change the function return type to _string_ and update consumers appropriately.

[the remark this answers is not in the document]
Comment (grandizzy) [reply pyexchange/okex.py]: sounds good!

File: pyexchange/okex.py [hunk 5]
```diff
@@ -217,118 +287,190 @@ def place_order(self, pair: str, is_sell: bool, price: Wad, amount: Wad) -> int:
...
 
-    def get_trades(self, pair: str, page_number: int = 1):
+    # Trading: Retrieves most recent 100 filled or partially filled orders for a pair.
+    def get_trades(self, pair: str, page_number: int = 1) -> List[Trade]:
         assert(isinstance(pair, str))
         assert(isinstance(page_number, int))
-        raise Exception("get_trades() not available for OKEX")
+        assert(page_number == 1)
+
+        result = self._http_get("/api/spot/v3/orders",
```
[the remark this answers is not in the document]
Comment (EdNoepel) [reply pyexchange/okex.py]: Letting urllib encode the statuses fixed that; thank you.

**Pull request merged**